_posts/contests/2021-02-05-Performance-analysis-and-tuning-contest-5.md

```diff
@@ -0,0 +1,216 @@
+
+## Couple of hints
+
+### The tools
+
+To measure the program's runtime, you can prefix the command with `time` or `multitime`, like this:
+
+```bash
+multitime -n 5 ../online2bin/online2-wav-nnet3-latgen-faster --word-symbol-table=graph/words.txt --config=conf/model.conf am/final.mdl graph/HCLG.fst ark:test.utt2spk scp:test.scp ark:/dev/null
+```
+
+You compare the runtime of your modification against the runtime of the original code. The faster your program, the better you are at optimizing.
+
+To find the hotspots (i.e. functions that take the most time), you will need to use the profiler. There are many profilers available: `perf` which is very common on Linux or Intel's VTune which works on Intel's chips only. Intel's VTune is a GUI, it is very useful and convenient both for beginners and experts. Alternatively, you can use `perf` in combination with `speedscope`, as described [here](https://johnysswlab.com/speedscope-visualize-what-your-program-is-doing-and-where-it-is-spending-time/).
+
+After having found the functions that are performance bottlenecks, the fun starts. One part of your work is to make sure that the CPU doesn't do unnecessary work, or think of ways to skip unnecessary work. This automatically translates to better performance since the number of instructions is lower.
+
+Another important way of making code run faster is to perform CPU-specific optimizations. The profiler tells you that the function is slow, but it doesn't tell you why. Are these data cache misses, branch prediction misses, instruction cache misses? To find out, you will need to use a special profiler that can read hardware performance counters and tell you that.
+
+Profilers that can read this kind of data are CPU-specific since each vendor and CPU family has a different set of counters. On Intel, you can use Intel's VTune Microarchitectural Analysis to get this information. Also, `pmu-tools` is another alternative (excellent information in Denis' [book](https://book.easyperf.net/perf_book)). On AMD, there is a profiler called uProf.
+
+Alternatively, you can use `likwid-perfctr` to read information from the hardware performance timers. This tool is open source and works with many types of CPUs from different vendors. Link to more information about `likwid-perfctr` is [here](https://johnysswlab.com/hardware-performance-counters-the-easy-way-quickstart-likwid-perfctr/).
+
+### What to measure?
```

Comment:
Do you think this will really add value? That's great info, but to me, it starts looking like a general educational blog post, not a contest announcement :) If it will become too long, folks will just skip reading or even worse, fall off from taking a challenge.